[core][1eventx/03] job event: send job events to the aggregator

can-anyscale opened the pull request:
This is part of a series of PRs to support JobEvent in the oneevent framework. The full effort will include adding the JobEvent schema, introducing a generic interface for exporting different types of events to the Event Aggregator, and implementing the necessary integration logic.

----

In this PR, we implement:
- The integration of RayEventRecorder to the GcsServer to export job event
- Add an e2e test to check that the expected job event is exported

Test:
- CI

Commit:
event recorder interface

Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
event recorder interface

Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
[core][1eventx/01] add protobuf for driver job event

Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
[core][1eventx/01] add protobuf for driver job event

Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
[core][1eventx/01] add protobuf for driver job event

Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
[core][1eventx/01] add protobuf for driver job event

Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
[core][1eventx/03] job event: export to Aggregator process

Signed-off-by: Cuong Nguyen <can@anyscale.com>

can-anyscale commented on src/ray/gcs/gcs_server/gcs_job_manager.cc:
```diff
@@ -124,7 +124,14 @@ void GcsJobManager::HandleAddJob(rpc::AddJobRequest request,
+
+    // Send driver job event to export API.
+    // TODO(can-anyscale): to be deprecated once the event aggregator is stable.
     WriteDriverJobExportEvent(job_table_data);
```
i should, good call out ;)

can-anyscale replied:
added

edoakes commented on python/ray/dashboard/modules/aggregator/tests/test_ray_job_events.py:
```diff
@@ -0,0 +1,56 @@
+    cluster.wait_for_nodes()
+    ray.init(address=cluster.address)
+
+    # Wait for the dashboard agent to be available
+    ip, _ = cluster.webui_url.split(":")
+    agent_address = f"{ip}:{ray_constants.DEFAULT_DASHBOARD_AGENT_LISTEN_PORT}"
+    assert wait_until_server_available(agent_address)
```
we should remove this from the test, it could mask the fact that the system loses data during the startup sequence

edoakes replied:
hm, I didn't realize that PR is related to events too, it only mentions metrics

edoakes commented on python/ray/dashboard/modules/aggregator/tests/test_ray_job_events.py:
```diff
@@ -0,0 +1,56 @@
+    assert req_json[0]["message"] == "driver job event"
+    assert (
+        req_json[0]["driverJobEvent"]["jobId"]
+        == base64.b64encode(b"\x01\x00\x00\x00").decode()
```
please don't use magic string. translate it from `ray.get_runtime_context().get_job_id()`

can-anyscale replied:
#fixed

edoakes commented on src/ray/gcs/gcs_server/BUILD.bazel:
```diff
@@ -197,6 +197,7 @@ ray_cc_library(
         "//src/ray/rpc:core_worker_client",
         "//src/ray/rpc:gcs_server",
         "//src/ray/stats:stats_metric",
+        "//src/ray/telemetry:ray_job_event_recorder",
```
the job manager should depend on a `job_event_recorder_interface`, not the concrete implementation

only the "main" file (in this case, `gcs_server.cc`) should depend on the concrete implementation

edoakes commented on src/ray/gcs/gcs_server/gcs_job_manager.cc:
```diff
@@ -124,7 +124,14 @@ void GcsJobManager::HandleAddJob(rpc::AddJobRequest request,
       running_job_start_times_.insert({job_id, job_table_data.start_time()});
     }
+
+    // Send driver job event to export API.
+    // TODO(can-anyscale): to be deprecated once the event aggregator is stable.
     WriteDriverJobExportEvent(job_table_data);
```
I would wrap this in a utility function and inside the function leave a comment explaining the context (that we're mid migration and `WriteDriverJobExportEvent` will be deprecated)

maybe `WriteDriverJobExportEvent` is already that utility function?

edoakes commented on src/ray/gcs/gcs_server/gcs_server.cc:
```diff
@@ -108,6 +108,10 @@ GcsServer::GcsServer(const ray::gcs::GcsServerConfig &config,
       }),
+      event_aggregator_client_(std::make_unique<rpc::EventAggregatorClientImpl>(
+          config_.metrics_agent_port, client_call_manager_)),
+      ray_job_event_recorder_(std::make_unique<telemetry::RayJobEventRecorder>(
+          *event_aggregator_client_, io_context_provider_.GetDefaultIOContext())),
```
what's the difference between an event "recorder" and an event "aggregator"

it seems like the "recorder" is just a client?

can-anyscale committed:
[core][1eventx/04] job event: use a separated thread (#55395)

This is part of a series of PRs to support JobEvent in the oneevent
framework. The full effort will include adding the JobEvent schema,
introducing a generic interface for exporting different types of events
to the Event Aggregator, and implementing the necessary integration
logic.

------------------

In this PR, we improve the isolation of the event exporting job from
other GCS components by using separate threads for periodic execution
and gRPC request handling.

Test:
- CI

---------

Signed-off-by: Cuong Nguyen <can@anyscale.com>

can-anyscale commented on src/ray/gcs/gcs_server/gcs_server.cc:
```diff
@@ -280,6 +284,9 @@ void GcsServer::DoStart(const GcsInitData &gcs_init_data) {
   InitUsageStatsClient();
 
+  // Start ray event recorder.
+  ray_job_event_recorder_->StartExportingEvents();
```
because of the lagging between GCS and DashboardAgent start up sequence; this needs to be invoked after DashboardAgent is ready (that "waiting" logic still needs to be implemented)

MengjinYan commented on BUILD.bazel:
```diff
@@ -144,6 +144,7 @@ ray_cc_library(
             "src/mock/ray/common/pubsub/publisher.h",
             "src/mock/ray/common/pubsub/subscriber.h",
             "src/mock/ray/common/ray_syncer/ray_syncer.h",
+            "src/mock/ray/observability/ray_ever_recorder_interface.h",
```
```suggestion
            "src/mock/ray/observability/ray_event_recorder_interface.h",
```

can-anyscale replied:
lol

MengjinYan commented on src/mock/ray/observability/ray_event_recorder_interface.h:
My understanding is we are trying to get rid of the mock classes and use dependency injection instead.

can-anyscale replied:
oh, is this something you're doing @edoakes? what is the new way of doing dependency injection?

edoakes replied:
ah yes please prefer to write a fake implementation instead of introducing mocks. the fake implementation should be shared across the codebase and can start simple but will evolve as it's used in more tests

you can see `FakeRayletClient` for an example

MengjinYan commented on src/ray/gcs/gcs_server/gcs_job_manager.cc:
```diff
@@ -46,18 +48,33 @@ void GcsJobManager::Initialize(const GcsInitData &gcs_init_data) {
     // don't write export event.
     return;
   }
+  if (RayConfig::instance().enable_ray_event()) {
```
As we are adding more events, probably we should consider adding config to control which type of events to export as ray event.

can-anyscale commented:
@MengjinYan's comments
